cpp/src/jni/dataset/jni_wrapper.cpp

```diff
@@ -0,0 +1,577 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <arrow/dataset/file_base.h>
+#include <arrow/dataset/api.h>
+#include <arrow/filesystem/localfs.h>
+#include <arrow/ipc/api.h>
+#include <arrow/util/iterator.h>
+#include <arrow/filesystem/hdfs.h>
+#include <arrow/io/api.h>
+#include <google/protobuf/io/coded_stream.h>
+#include <google/protobuf/message.h>
+#include "concurrent_map.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/cast.h"
+#include "arrow/compute/kernels/compare.h"
+#include "jni/dataset/Types.pb.h"
+
+#include "org_apache_arrow_dataset_jni_JniWrapper.h"
+#include "org_apache_arrow_dataset_file_JniWrapper.h"
+
+static jclass illegal_access_exception_class;
+static jclass illegal_argument_exception_class;
+static jclass runtime_exception_class;
+
+static jclass record_batch_handle_class;
+static jclass record_batch_handle_field_class;
+static jclass record_batch_handle_buffer_class;
+
+static jmethodID record_batch_handle_constructor;
+static jmethodID record_batch_handle_field_constructor;
+static jmethodID record_batch_handle_buffer_constructor;
+
+static jint JNI_VERSION = JNI_VERSION_1_6;
+
+using arrow::jni::ConcurrentMap;
+
+static ConcurrentMap<std::shared_ptr<arrow::dataset::DatasetFactory>> dataset_factory_holder_;
```

Comment:
this still strikes me as a bad pattern.

https://stackoverflow.com/questions/26375215/c-shared-ptr-and-java-native-object-ownership/26377515 seems better.

Reply:
Will do. But I kindly like the current pattern - it benefits more or less when debugging/tracing (self-growing id, instance cache, etc.), right? Is there any other consideration making it bad?

Reply:
done

Reply:
extra memory and resolution overhead.  The instance cache should be handled by the underlying allocator which should be highly optimized for returning pointer sized values.

Could you explain how a self-growing ID is useful for debugging purposes?  If it is required would a DEBUG log with only an atomic int associated with each new ID be sufficient?

Reply:
> Could you explain how a self-growing ID is useful for debugging purposes?

It may not be a considerable bit deal but growing ID numbers can be much smaller and more contextual. E.g. when we experience a JVM breakpoint debugging against some sort of abnormal duplicated buffer creation, if five different buffers with ID 10-15 are created one would easily remember it. Then if ID 12 is returned from JNI again then one instantly realizes that something must went wrong there.

> If it is required would a DEBUG log with only an atomic int associated with each new ID be sufficient?

Yes I think if we are debugging with logs rather than breakpoints, what type of ID is used is indeed not that important because we can use `control + F` to highlight a specific instance anyway.

And as the performance impact is considered I don't have strong inclination. So let's just use pointers?

Another thing is we already use maps to cache JNI objects in other modules (gandiva-jni, orc-jni), we may need to unify the approaches somehow.